crates/storage/provider/src/providers/static_file/manager.rs

```diff
@@ -448,6 +452,161 @@ impl StaticFileProvider {
+            let new_target = new_target.unwrap_or_default();
+            if let Some(target) = unwind_target.as_mut() {
+                *target = (*target).min(new_target);
+            } else {
+                unwind_target = Some(new_target);
+            }
+        };
+
+        for segment in StaticFileSegment::iter() {
+            if has_receipt_pruning && segment.is_receipts() {
+                // Pruned nodes (including full node) do not store receipts as static files.
+                continue
+            }
+
+            let initial_highest_block = self.get_highest_static_file_block(segment);
+
+            // By calling a writer, an inner consistency check will be triggered on
+            // [`NippyJarWriter::check_consistency_and_heal`]
+            //
+            // * If appending data was interrupted before a config commit, then data file will be
+            //   truncated according to the config.
+            //
+            // * If pruning data was interrupted before a config commit, then we have deleted data
+            //   that we are expected to still have. We need to check the Database and unwind
+            //   everything accordingly.
+            let mut writer = self.latest_writer(segment)?;
+
+            // This will commit any inconsistencies found and update the reader index. If it
+            // happens, it means that the highest block of the segement has decreased.
+            writer.commit()?;
+
+            // Only applies to block-based static files. (Headers)
+            //
+            // The updated `highest_block` may have decreased if we healed from a pruning
+            // interruption.
+            let mut highest_block = self.get_highest_static_file_block(segment);
+            if initial_highest_block != highest_block {
+                update_unwind_target(highest_block);
+            }
+
+            // Only applies to transaction-based static files. (Receipts & Transactions)
+            //
+            // Make sure the last transaction matches the last block from its indices, since a heal
+            // from a pruning interruption might have decreased the number of transactions without
+            // being able to update the last block of the static file segment.
+            let highest_tx = self.get_highest_static_file_tx(segment);
+            if let Some(highest_tx) = highest_tx {
+                let mut last_block = highest_block.unwrap_or_default();
+                loop {
+                    if let Some(indices) = provider.block_body_indices(last_block)? {
+                        if indices.last_tx_num() <= highest_tx {
+                            break
+                        }
+                    }
+                    if last_block == 0 {
+                        break
+                    }
+                    last_block -= 1;
+
+                    highest_block = Some(last_block);
+                    update_unwind_target(highest_block);
+                }
```

Comment:
can we move this to static file manager, so that we can re-use it in stages as well?